code/iaas/cluster-logic/src/main/java/io/cattle/iaas/cluster/service/impl/ClusterManagerImpl.java

```diff
@@ -235,33 +263,397 @@ public boolean evaluate(Instance obj) {
+                    throw e;
+                } catch (Exception e) {
+                    throw new RuntimeException(e);
+                }
+            }
+            deploySslAgent(hostInCluster);
+            updateSslConfig(state, hostInCluster);
+        }
+    }
+
+    @SuppressWarnings("deprecation")
+    private void createDockerCertificates(Host host) throws Exception {
+        Certificate caCertRecord = getOrCreateCaCertificates(host.getAccountId());
+
+        Certificate serverCertRecord;
+        if (host.getCertificateId() != null) {
+            serverCertRecord = objectManager.loadResource(Certificate.class, host.getCertificateId());
+            if (serverCertRecord != null &&
+                    serverCertRecord.getCertChain() != null &&
+                    serverCertRecord.getCertChain().equals(caCertRecord.getCert())) {
+                // docker ssl cert for host currently exist and matches the current certificate
+                // authority
+                return;
+            }
+        }
+
+        KeyPair caKeyPair = SshKeyGen.readKeyPair(caCertRecord.getKey());
+
+        KeyPair serverKeyPair = SshKeyGen.generateKeyPair();
+        String serverKey = SshKeyGen.writePemObject(serverKeyPair.getPrivate());
+
+        IpAddress ipAddress = clusterHostMapDao.getIpAddressForHost(host.getId());
+
+        PKCS10CertificationRequest serverCSR = new PKCS10CertificationRequest(
+                "SHA256withRSA",
+                new X500Principal("CN="),
+                serverKeyPair.getPublic(),
+                null,
+                serverKeyPair.getPrivate());
+        X509V3CertificateGenerator serverCertGen = new X509V3CertificateGenerator();
+        serverCertGen.setSerialNumber(BigInteger.valueOf(System.currentTimeMillis()));
+        serverCertGen.setIssuerDN(CertificateConstants.CERT_ISSUER);
+        serverCertGen.setNotBefore(new Date());
+        serverCertGen.setNotAfter(new Date(System.currentTimeMillis() + CertificateConstants.EXPIRATION_MILLIS.get()));
+        serverCertGen.setSubjectDN(serverCSR.getCertificationRequestInfo().getSubject());
+        serverCertGen.setPublicKey(serverCSR.getPublicKey());
+        serverCertGen.setSignatureAlgorithm("SHA1WithRSAEncryption");
+
+        serverCertGen.addExtension(X509Extensions.ExtendedKeyUsage, false, new ExtendedKeyUsage(
+                KeyPurposeId.id_kp_serverAuth));
+        GeneralNames subjectAltName = new GeneralNames(new DERSequence(new GeneralName[] {
+                new GeneralName(GeneralName.iPAddress, ipAddress.getAddress()),
+                new GeneralName(GeneralName.iPAddress, "127.0.0.1")
+        }));
+        serverCertGen.addExtension(X509Extensions.SubjectAlternativeName, false, subjectAltName);
+
+        X509Certificate serverCrt = serverCertGen.generateX509Certificate(caKeyPair.getPrivate());
+        String serverCert = SshKeyGen.writePemObject(serverCrt);
+
+        // create cert record
```

Comment:
Would it make sense to wrap 2 updates below in a single transaction? so certificate gets created and the host gets updated with the cert in a transaction with the single point of failure